File: frontend/src/services/wallet.ts
/**
 * Wallet API Service
 */
import api from './api';

export interface WalletJournalEntry {
  id: number;
  entry_id: number;
  date: string;
  ref_type: string;
  amount: number;
  balance: number | null;
  description: string;
  first_party_id: number | null;
  second_party_id: number | null;
}

export interface WalletTransaction {
  id: number;
  transaction_id: number;
  date: string;
  type_id: number;
  quantity: number;
  unit_price: number;
  is_buy: boolean;
  client_id: number;
  location_id: number;
}

export interface WalletStatistics {
  period_days: number;
  total_income: number;
  total_expenses: number;
  net_change: number;
  market_buys: number;
  market_sells: number;
  market_profit: number;
}

export const walletService = {
  /**
   * Get wallet balance
   */
  getBalance: async (characterId: number) => {
    const response = await api.get(`/wallet/balance/${characterId}`);
    return response.data;
  },

  /**
   * List wallet journal entries
   */
  listJournal: async (params?: {
    character_id?: number;
    from_date?: string;
    to_date?: string;
    ref_type?: string;
    limit?: number;
    offset?: number;
  }) => {
    const response = await api.get<WalletJournalEntry[]>('/wallet/journal/', { params });
    return response.data;
  },

  /**
   * List wallet transactions
   */
  listTransactions: async (params?: {
    character_id?: number;
    from_date?: string;
    to_date?: string;
    type_id?: number;
    is_buy?: boolean;
    limit?: number;
    offset?: number;
  }) => {
    const response = await api.get<WalletTransaction[]>('/wallet/transactions/', { params });
    return response.data;
  },

  /**
   * Get wallet statistics
   */
  getStatistics: async (characterId: number, days: number = 30) => {
    const response = await api.get<WalletStatistics>(`/wallet/statistics/${characterId}`, {
      params: { days },
    });
    return response.data;
  },

  /**
   * Trigger wallet sync
   */
  triggerSync: async (characterId: number) => {
    const response = await api.post(`/wallet/sync/${characterId}`);
    return response.data;
  },
};

File: backend/app/models/wallet.py
"""
Wallet models

EVE Online wallet transactions and journal
"""
from sqlalchemy import Column, Integer, String, DateTime, Float, ForeignKey, BigInteger, Numeric, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
from app.core.database import Base


class WalletJournal(Base):
    """Character or corporation wallet journal entries"""

    __tablename__ = "wallet_journal"

    id = Column(Integer, primary_key=True, index=True)
    character_id = Column(Integer, ForeignKey("characters.id"), nullable=True, index=True)
    corporation_id = Column(Integer, ForeignKey("corporations.id"), nullable=True, index=True)

    # Journal entry data
    entry_id = Column(BigInteger, unique=True, index=True)
    date = Column(DateTime(timezone=True), index=True)
    ref_type = Column(String(100), index=True)  # Transaction type (bounty_prizes, market_escrow, etc.)
    amount = Column(Numeric(precision=20, scale=2))
    balance = Column(Numeric(precision=20, scale=2), nullable=True)
    description = Column(Text)

    # Context IDs
    context_id = Column(BigInteger, nullable=True)  # Related entity ID
    context_id_type = Column(String(50), nullable=True)  # Type of context_id (contract_id, market_transaction_id, etc.)

    # First and second party
    first_party_id = Column(BigInteger, nullable=True)
    second_party_id = Column(BigInteger, nullable=True)

    # Tax information
    tax = Column(Numeric(precision=20, scale=2), nullable=True)
    tax_receiver_id = Column(BigInteger, nullable=True)

    # Reason (optional description from user)
    reason = Column(String(500), nullable=True)

    # Metadata
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    # Relationships
    character = relationship("Character", back_populates="wallet_journal")

    def __repr__(self):
        return f"<WalletJournal(entry_id={self.entry_id}, ref_type='{self.ref_type}', amount={self.amount})>"


class WalletTransaction(Base):
    """Character or corporation wallet market transactions"""

    __tablename__ = "wallet_transactions"

    id = Column(Integer, primary_key=True, index=True)
    character_id = Column(Integer, ForeignKey("characters.id"), nullable=True, index=True)
    corporation_id = Column(Integer, ForeignKey("corporations.id"), nullable=True, index=True)

    # Transaction data
    transaction_id = Column(BigInteger, unique=True, index=True)
    date = Column(DateTime(timezone=True), index=True)
    type_id = Column(Integer, index=True)  # Item type ID
    quantity = Column(BigInteger)
    unit_price = Column(Numeric(precision=20, scale=2))
    is_buy = Column(String(20), index=True)  # True = buy, False = sell

    # Client and location
    client_id = Column(BigInteger)
    location_id = Column(BigInteger)
    is_personal = Column(String(20), default=True)
    journal_ref_id = Column(BigInteger, nullable=True)  # Link to journal entry

    # Metadata
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    # Relationships
    character = relationship("Character", back_populates="wallet_transactions")

    def __repr__(self):
        return f"<WalletTransaction(transaction_id={self.transaction_id}, type_id={self.type_id}, quantity={self.quantity})>"
